tokio/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+//!         // Register with the Tokio runtime here, e.g.:
+//!         // `tokio_cmp.runtime()?.spawn(async move { ... });`
+//!         Ok(())
+//!     }
+//! }
+//! ```
+//!
+//! ## Add `TokioComponent` to your Abscissa application
+//!
+//! Inside of your app's `application.rs`, find the [`register_components`]
+//! method and add [`TokioComponent`]:
+//!
+//! ```text
+//! fn register_components(&mut self, command: &Self::Cmd) -> Result<(), FrameworkError> {
+//!     let mut components = self.framework_components(command)?;
+//!
+//!     // Create `TokioComponent` and add it to your app's components here:
+//!     use abscissa_tokio::TokioComponent;
+//!     components.push(Box::new(TokioComponent::new()?));
+//!
+//!     self.state.components.register(components)
+//! }
+//! ```
+//!
+//! Inside of the [`Runnable`] for one of your application's subcommands, call
+//! [`abscissa_tokio::start`] to launch the Tokio runtime:
+//!
+//! ```text
+//! impl Runnable for StartCmd {
+//!    fn run(&self) {
+//!        abscissa_tokio::start(&crate::application::APPLICATION);
+//!    }
+//! }
+//! ```
+//!
+//! This will run any futures which were registered
+//!
+//! [Tokio]: https://tokio.rs
+//! [`tokio::main`]: https://docs.rs/tokio/latest/tokio/attr.main.html
+//! [`abscissa_core::boot`]: https://docs.rs/abscissa_core/latest/abscissa_core/application/fn.boot.html
+//! [`Component`]: https://docs.rs/abscissa_core/latest/abscissa_core/component/trait.Component.html
+//! [`TokioComponent`]: https://docs.rs/abscissa_tokio/latest/abscissa_tokio/struct.TokioComponent.html
+//! [`register_components`]: https://docs.rs/abscissa_core/latest/abscissa_core/application/trait.Application.html#tymethod.register_components
+//! [`Runnable`]: https://docs.rs/abscissa_core/latest/abscissa_core/trait.Runnable.html
+//! [`abscissa_tokio::start`]: https://docs.rs/abscissa_tokio/latest/abscissa_tokio/application/fn.start.html
+
+#![doc(
+    html_logo_url = "https://www.iqlusion.io/img/github/iqlusioninc/abscissa/abscissa-sq.svg",
+    html_root_url = "https://docs.rs/abscissa_tokio/0.5.0-pre"
+)]
+#![forbid(unsafe_code)]
+#![warn(rust_2018_idioms, unused_lifetimes, unused_qualifications)]
+
+pub use tokio;
+
+use abscissa_core::{
+    application::{AppCell, Application},
+    format_err, Component, FrameworkError, FrameworkErrorKind,
+};
+use futures_util::future;
+use tokio::runtime::Runtime;
+
+/// Start the Tokio runtime within the given application.
+///
+/// Panics if the runtime is unavailable or has already been started.
+pub fn start<A: Application>(app_cell: &'static AppCell<A>) {
+    let mut runtime = app_cell
+        .write()
+        .state_mut()
+        .components
+        .get_downcast_mut::<TokioComponent>()
+        .expect("TokioComponent not registered!")
+        .runtime
+        .take()
+        .expect("Tokio runtime has already been started!");
+
+    runtime.block_on(future::pending::<()>())
```

Comment:
How does the application shut down in this model?

Reply:
I had assumed this would return after running all futures to completion. Is that not the case? (pretty much anything we’re going to use this for is designed to run indefinitely, so I hadn’t tested that assumption)

Reply:
Nope, `future::pending` never resolves, so this will never terminate.  Even for stuff that's designed to run indefinitely, it seems like it might be useful to have a shutdown mechanism short of killing the application.

Maybe a better design would be for the component to create an mpsc channel of type `()` and await the receiver in `runtime.block_on`, so that sending a message through the channel shuts down the application.  The sender end of the channel could then be accessed through some method on the component.